js/runner.ts

```diff
@@ -0,0 +1,187 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+// tslint:disable-next-line:no-circular-imports
+import * as deno from "./deno";
+import { globalEval } from "./global_eval";
+import { assert, log } from "./util";
+
+// tslint:disable:no-any
+type AmdCallback = (...args: any[]) => void;
+type AmdDefine = (deps: ModuleSpecifier[], factory: AmdFactory) => void;
+type AmdErrback = (err: any) => void;
+type AmdFactory = (...args: any[]) => object | void;
+type AmdRequire = (
+  deps: ModuleSpecifier[],
+  callback: AmdCallback,
+  errback: AmdErrback
+) => void;
+// tslint:enable:no-any
+
+// tslint:disable-next-line:no-any
+type BuiltinMap = { [moduleSpecifier: string]: any };
+
+// Type aliases to make the code more readable
+type ContainingFile = string;
+type Filename = string;
+type ModuleSpecifier = string;
+type OutputCode = string;
+
+/** Internal representation of a module being loaded */
+class Module {
+  deps?: Filename[];
+  factory?: AmdFactory;
+
+  // tslint:disable-next-line:no-any
+  constructor(public filename: Filename, public exports: any) {}
+}
+
+/** External APIs which the runner depends upon to be able to retrieve
+ * modules.
+ */
+export interface Provider {
```

Comment:
The name "Provider" could be more specific... "CodeProvider"? "CodeLoader"?

Reply:
I like `CodeProvider`